include/zephyr/pm/power_gating.h

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2022 Intel corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_POWER_GATING_H_
+#define ZEPHYR_INCLUDE_POWER_GATING_H_
+
+#include <kernel.h>
+
+/**
+ * @brief Power gating API
+ * @defgroup subsys_pm_pg Power Gating
+ * @ingroup subsys_pm
+ */
+
+/**
+ * @struct power_reg
+ * @brief Structure for specifying address and bits used in power gating
+ *
+ * @param wr_address Address of register used to set PG
+ * @param rd_address Address of register used to read PG status
+ * @param SPA_bit Set Power Active bit in wr_address used to toggle power on/off
+ * @param CPA_bit Confirm Power Active bit in rd_address used to verify power gating state
+ */
+struct pm_pg_registers {
+  uint32_t wr_address;
+	uint32_t rd_address;
+  uint32_t SPA_bit;
+  uint32_t CPA_bit;
+};
+
+/**
+ * @brief Set device power on/off
+ * 
+ * Function will set reg SPA_bit in wr_address and wait for
+ * CPA_bit in rd_address to change. Will return -EIO on reaching us timoeut.
+ *
+ * @param reg pm_pg_registers specifying devices pg registers
+ * @param power_enable false to turn device off, true to turn on
+ * @param us CPA timoeut in us
+ */
+int device_set_power_enable(struct pm_pg_registers reg, bool power_enable, uint64_t us);
```

Comment:
```suggestion
int device_set_power_enable(struct pm_pg_registers reg, bool power_enable, uint64_t us);
```
I think that if this is going to be an API for this device driver, it should at least notify the the device that it is being powered on. Just like it is being done in `ssp_pm_action`

Reply:
As above, this is not supposed to be a stand alone turn on/off mechanism - just a part common for all intel devices. If it's too problematic, we can drop and make every device implement it by itself.

Reply:
Not understanding your point. I got this is one on/off mechanism, but there are devices that will be affected when you trigger it right ? Or am I missing how this is supposed to be used ? Who is supposed to call it ?

Reply:
The general / recommended way to interact with a device PM is using the device runtime API. Having something like this is problematic because it is not integrated in Zephyr's device model. For example, from Zephyr's perspective this device will look as ACTIVE even if you call `device_set_power_enable(XXX, false, YYY);`.

Is `reg` controlling other devices power ? If so, I strongly recommend each device do it by itself in their pm action callback.

Reply:
Have a look on the usage in SSP. I've deleted the `dai_ssp_pm_runtime_en/dis_ssp_power` and replaced them by 'pg_set_device_power_enable'. Reasoning is that there's no point in reimplementing this mechanism multiple times, when it's the same for multiple devices (SSP, DMIC, SNDW, UART, SPI, etc.)
It should be called from inside of a device, passing it's specific register/bit data. 
So for ssp, `ssp_pm_action `(PM_DEVICE action) during `PM_DEVICE_ACTION_RESUME` calls for `dai_ssp_probe`. This does all of the necessary turn-on procedures: clock gating, emptying buffers and calls for `pg_set_device_power_enable`. 
And for DMIC - which does not yet use zephyr power management - it would replace `dai_dmic_en/dis_power`. (see [dai.c](https://github.com/zephyrproject-rtos/zephyr/pull/45563/files#diff-73640fcfea556c10ea0ffd5709a24f77de2152e5df84109ad2512c511909cfeb))
Only one device per function call is affected - the one which register/bit set is passed to it. 
It does not invalidate nor affect the usage of zephyr power management system.
It might be problematic if used outside of a devices turn on/off sequence (which is wrong and not intended at all). Do you have an idea on how to guard from this?

Reply:
No, that is not how it works on Zephyr. You are introducing a new API that does not fit well on Zephyr's power management infra-struct. Zephyr's power management interacts with devices if you don't implement pm per device it will not work.

Reply:
@ceolin So I'd suggest discarding this problematic part for now, focus on getting the power domains part and running adsp devices with zephyr pm. This does not require any additional apis. Than we can have a talk on what/if anything can be done with the reoccurring parts of code. Is it ok with you?